File: Prob2_MergeSortedArrays.java

// Time Complexity : O(m + n)
// Space Complexity : O(1)

// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
/*
Maintaing 3 pointers : One for traversing from last element != 0 of nums1, last element of nums2 and
pointer for merging the arrays, which is pointed at last element in nums1

WHICHEVER element is greater between nums1 and nums2, will be copied at pointer we used for merge.
Till pointer is >= 0

IF some elements are left to copy from nums2 to nums2
We will simply copy one by one till all the elements are copied into nums1 


*/

// Your code here along with comments explaining your approach


class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        if(nums1 == null)   return;
        
        int ptr1 = m - 1, ptr2 = n - 1, i = m + n - 1;
        
        while(ptr1 >= 0 && ptr2 >= 0){
            
            if(nums1[ptr1] > nums2[ptr2]){
                nums1[i] = nums1[ptr1];
                ptr1--;
            }else{
                nums1[i] = nums2[ptr2];
                ptr2--;
            }
            i--;
        }
        
        while(ptr2 >= 0){
            nums1[i] = nums2[ptr2];
            i--; ptr2--;
        }
    }
}

File: prob-2-merge-sorted-arrays.py
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Time complexity: O(n+m)
        Space complexity: O(m) - bigger array
        Solution: We will use the 3-pointers solution from the beginning to solve this
        Did this code successfully run on Leetcode: Yes
        Any problem you faced while coding this: No
        """

        nums1_copy = nums1[:m]

        p1 = 0 # pointer for nums1_copy
        p2 = 0 # pointer for nums2

        # Compare elements from nums1_copy and nums2 
        # Copy smallest to num1
        for p in range(m+n):
            # Check that p1 and p2 do not cross the boundaries
            if (p2 >= n) or (p1 < m and nums1_copy[p1] < nums2[p2]):
                 nums1[p] = nums1_copy[p1]
                 p1 += 1
            else:
                nums1[p] = nums2[p2]
                p2 += 1



        
